### PR TITLE
[Marian Fixes] prevent predicting pad_token_id before softmax, support language codes, name multilingual models

### LINKED ISSUE #3775: OpusNMT/MarianMT Machine Translation Models
### Model description
1,026 Language Pair Models, downloadable [here](http://opus.nlpl.eu/Opus-MT/)
Trained with Marian C++ [library](https://github.com/marian-nmt/marian)

### Open source status

* [ x] the model implementation is available: (give details)
* [ x] the model weights are available: (give details)
* [ x] who are the authors: TODO, find gh-usernames of authors!

### Proposed API:
```python
model_name = 'marian/en-fr'
tokenizer = MarianTokenizer.from_pretrained(model_name)
model = MarianModel.from_pretrained(model_name)
src_text = "how are you today?"
tgt_text = "comment allez vous aujourd'hui?"
# Training API
full_inputs: dict = tokenizer.prepare_batch_for_translation(src_text, tgt_text=tgt_text)
loss, logits, *other_outputs = model(
        full_inputs['input_ids'],
        full_inputs['attention_mask'],

        full_inputs['decoder_input_ids'], # this argument is mandatory for the forward pass
)
# Inference/generate API
src_inputs: dict = tokenizer.prepare_batch_for_translation(src_text)
generated_fr_ids = model.generate(src_inputs['input_ids'], src_inputs['attention_mask'])
french_text: List[str] = processor.decode_batch(generated_fr_ids)
```

### Implementation Details
`MarianTokenizer` Signatures

(Not calling it Tokenizer to avoid confusion, but I don't feel strongly.)
- All models require `MosesSentenceSplitter` and `MosesPunctuationNormalizer` preprocessing
    -  There are some additional perl scripts we will not port for pre/post-processing
- 81 of the models require BPE, 960 require SentencePiece.
- We can decide which is which

```python
class MarianTokenizer:
    def __init__(self, vocab_file, source_bpe, target_bpe, source_spm, target_spm):
        # decide whether to use BPE/SPM based on which files are present in S3.
        self.source_lang, self.target_lang #inferred from paths/config
        #self.source_spm =

    @property
    def uses_sentencepiece(self) -> bool:
        return self.source_spm is not None

    def from_pretrained(self, *args, **kwargs):
        # needs to be overwritten or modified to not fail if certain files not present.

    def prepare_batch_for_translation(src_text:str, tgt_text=None, return_tensors='pt', max_length=512, pad_to_max_length=True) -> Dict[str, tensor/list]:
		return {}


    def decode_batch(self, target_lang_ids: List[List[int]]) -> List[str]:


    def decode(self, target_lang_id: List[int]) -> str:
```

#### Edits
- renamed `MarianProcessor` -> `MarianTokenizer` for consistency

### LINKED ISSUE #4160: [Marian] Multilingual models require language codes
[ ] figure out codes
[ ] update/subclass `MarianTokenizer`

### PR DESCRIPTION
**Pull request by sshleifer**
(this should probably be 3 PRs, sorry).

TLDR: This gets us to equal or better BLEU for 99+% of the converted models.

### Issue 1: non-breaking Generation Issue
`pad_token_id` is predicted with score 0, but we never want to generate it.
Previously, we set its score to inf  after softmax, so probabilities summed to something much less than 1. At the end of long generations, the probabilities got infinitesimally small, and BLEU suffered.

Solution: set `logits[:, pad_token_id] = -float('inf')` before calling `log_softmax`
Also, renames `prepare_scores_for_generation` -> `prepare_logits_for_generation` because it is now called on logits.

- Bart outputs are unchanged, other models unaffected.
- This could also be done for other filters, like bad_words_ids, but results may change slightly.
- This works fine with fp16
- Improves BLEU scores for 50% of Marian models, often by a huge amount for those with long examples in test set. No change for others models.

### Issue 2 Multilingual Models require language codes.

`>>pt<< Don't spend so much time watching TV.`
The language code at the beginning, which tells the model to generate in portuguese, is not known to sentencepiece, so `MarianTokenizer` must manually extract it and add it as it's own piece.
We do this using `re.match` inside `_tokenize`.

### Issue 3: Opus' Multilingual models have very long names
We remap them to much shorter names, like `Helsinki-NLP/opus-mt-en-ROMANCE` in `convert_marian_to_pytorch.py`, and document the naming scheme in `docs/`.

### Working snippet:
```
from transformers import MarianMTModel, MarianTokenizer
text = ['>>fr<< this is a sentence in english that we want to translate to french', 
        '>>pt<< This should go to portuguese',
        '>>es<< And this to Spanish'
       ]
model_name = 'Helsinki-NLP/opus-mt-en-ROMANCE'
tokenizer = MarianTokenizer.from_pretrained(model_name)
model = MarianMTModel.from_pretrained(model_name)
translated = model.generate(**tokenizer.prepare_translation_batch(text))
[tokenizer.decode(t, skip_special_tokens=True) for t in translated]

=> ["c'est une phrase en anglais que nous voulons traduire en français",
 'Isto deve ir para o português.',
 'Y esto al español']
```

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
The slow bart summarization tests all pass with this?

**Comment by sshleifer** [comment]
yes!

**Comment by patrickvonplaten** [comment]
LGTM

**Review by patrickvonplaten — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
LGTM. I guess `prepare_scores_for_generation` should only be used as an internal and is therefore not a breaking change?

**Comment by sshleifer** [comment]
Correct, @LysandreJik not a breaking change.

File: src/transformers/modeling_bart.py [hunk 1]
```diff
@@ -980,12 +980,12 @@ def prepare_inputs_for_generation(self, decoder_input_ids, past, attention_mask,
             "use_cache": use_cache,  # change this to avoid caching (presumably for debugging)
         }
 
-    def prepare_scores_for_generation(self, scores, cur_len, max_length):
+    def prepare_logits_for_generation(self, logits, cur_len, max_length):
```
Comment (sshleifer) [inline src/transformers/modeling_bart.py:983]: could make this private

**Comment by sshleifer** [comment]
Merging in advance of release tomorrow. Feel free to leave comments and I will fix them!
